### PR TITLE
Add to_bson methods where missing. New tests and improve current tests.

### PR DESCRIPTION
**Pull request by guyboertje**
NOTES TO REVIEWERS:
- This gives our Event, Timestamp class and BigDecimal proper bson support like the BSON ruby library does.
- There are tests for the bson support.
- We are not currently using the Event bson support at the moment because we mutate the hash before giving it to Mongo
- Improved the `receive(event)` method a bit.
- Minor improvement to the mongo unit test
- major improvement to the the integration test. Used a uuid to search for the document in Mongo because the timestamp may be an ISODate object and not a String. Added a few other objects incl. BigDecimal to the Event content to better test integration.  Although the test does not check the contents of what it reads back from Mongo, I did it manually in the mongo console.
- I have a question: why do we omit the metadata from the document sent to Mongo?

From Mongo console

```
MongoDB shell version: 3.0.7
connecting to: test
Server has startup warnings:
2015-12-08T12:34:11.589+0000 I CONTROL  [initandlisten]
2015-12-08T12:34:11.589+0000 I CONTROL  [initandlisten] ** WARNING: soft rlimits too low. Number
  of files is 256, should be at least 1000
> use logstash
> db.logs.find()
{ "_id" : ObjectId("56671628146fbfeff9000001"), "message" : "This is a message!",
"uuid" : "3777a8f0-99a5-4df6-8772-d3e9392fd374", "number" : 4321.1234,
"utf8" : "żółć", "int" : 42, "arry" : [ 42, "string", 4321.1234 ],
"@version" : "1", "@timestamp" : ISODate("2015-12-08T17:40:56.932Z") }
```

### CODE REVIEW

**Comment by guyboertje** [comment]
We should back port this to 1.5

**Comment by purbon** [comment]
@guyboertje makes sense, we should backport. Will review this soon.

**Comment by purbon** [comment]
@guyboertje can you clarify me this `why do we omit the metadata from the document sent to Mongo?`? do you mean logstash metadata or mongo metadata?

File: lib/logstash/outputs/bson/big_decimal.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Copyright (C) 2009-2014 MongoDB Inc.
```
Comment (purbon) [inline lib/logstash/outputs/bson/big_decimal.rb:1]: do we need this copyright header information? not sure we have to as this is not modifying their code, right? what do you think?

**Comment by guyboertje** [comment]
@purbon - I mean why do we not include the Event metadata in the hash we use to create the Mongo document. What are the pros and cons. Should we have a config option?

**Comment by purbon** [comment]
@guyboertje at my knowledge, we don't store the event metadata anywhere, is intended only to life within the pipeline. what are you exactly thinking? can you elaborate a bit more your wish?

File: lib/logstash/outputs/bson/big_decimal.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  # to and from # raw bytes as specified by the BSON spec.
+  #
+  # @see http://bsonspec.org/#/specification
+  module BigDecimal
+
+    # A floating point is type 0x01 in the BSON spec.
+    BSON_TYPE = 1.chr.force_encoding(BINARY).freeze
+
+    # The pack directive is for 8 byte floating points.
+    PACK = "E".freeze
```
Comment (purbon) [inline lib/logstash/outputs/bson/big_decimal.rb:29]: is there any reason to use little-endian format here? I mean not saying is wrong, just asking to know why :smile_cat:

Comment (guyboertje) [inline lib/logstash/outputs/bson/big_decimal.rb:29]: Straight copy pasta from ruby-bson

File: lib/logstash/outputs/bson/big_decimal.rb (hunk 1 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/big_decimal.rb:1]: Well most of the code is pasted from ruby-bson.  I don't know the rules very well. So I put it in so we can discuss whether it should stay.  If it was not there no one would question the missing copyright text :-)

**Comment by guyboertje** [comment]
OK if metadata is internal, ignore my comment.  I will remove it.

File: lib/logstash/outputs/bson/logstash_event.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    # Get the event as encoded BSON.
+    # @example Get the hash as encoded BSON.
+    #   Event.new("field" => "value").to_bson
+    # @return [ String ] The encoded string.
+    # @see http://bsonspec.org/#/specification
+     def to_bson(buffer = ByteBuffer.new)
+      position = buffer.length
+      buffer.put_int32(0)
+      to_hash_with_metadata.each do |field, value|
```
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb]: why including metadata in the serialization? does it makes sense? I think we're not storing metadata anywhere, so is only a temporal thing living into the pipeline. Were you thinking about a use case here? I would love to know :+1:

File: lib/logstash/outputs/bson/logstash_event.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        buffer.put_byte(value.bson_type)
```
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb:37]: I wonder if there might be a situation where the value got no `bson_type`, does this makes sense?

File: lib/logstash/outputs/bson/logstash_event.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    end
+
+    module ClassMethods
+      # Deserialize the Event from BSON.
+      # @param [ ByteBuffer ] buffer The byte buffer.
+      # @return [ Event ] The decoded bson document.
+      # @see http://bsonspec.org/#/specification
+      def from_bson(buffer)
+        hash = Hash.new
+        buffer.get_int32 # Throw away the size.
```
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb:60]: If we throw it away? I wonder why do we need it? is this some kind of spec requirement?

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 3 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_event.rb]: We are not using `event.to_bson` at the moment, but I will change it to `to_hash` anyway - for the reasons you give.

File: lib/logstash/outputs/bson/logstash_event.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+# Modified 2015 Elastic
+
+module BSON
+
+  # Injects behaviour for encoding and decoding time values to
+  # and from raw bytes as specified by the BSON spec.
+  #
+  # @see http://bsonspec.org/#/specification
+  module LogStashEvent
```
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb:23]: What do you think? I would use two module declarations here insted of `LogStashEvent`, something like module LogStash; module Event. I find this more clear.

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 4 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_event.rb:37]: There might be a value we have in our event data that has not been extended for bson like BigDecimal and Timestamp have.  Its a bug that we need to fix when someone creates an issue about it.

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 5 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_event.rb:60]: Again, copied from bson hash extension.

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 6 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_event.rb:23]: No need, this module only lives to be included and extended in another class. See lines 74,75.

File: lib/logstash/outputs/bson/logstash_timestamp.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+# Modified 2015 Elastic
+
+module BSON
+
+  # Injects behaviour for encoding and decoding time values to
+  # and from raw bytes as specified by the BSON spec.
+  #
+  # @see http://bsonspec.org/#/specification
+  module LogStashTimestamp
```
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_timestamp.rb:23]: same here about module declaration as commented before.

Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_timestamp.rb:23]: See above comment.

File: lib/logstash/outputs/bson/big_decimal.rb (hunk 1 above)
Comment (purbon) [inline lib/logstash/outputs/bson/big_decimal.rb:1]: sure, I don't have either the knowledge to decide if should stay or can be removed. @suyograo @jordansissel can you chime in ? what do you think?

File: lib/logstash/outputs/bson/big_decimal.rb (hunk 2 above)
Comment (purbon) [inline lib/logstash/outputs/bson/big_decimal.rb:29]: ok, fair point. at the end of the day using little endian makes sense to me, was just wondering if there were any reasoning behind from you.

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 6 above)
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb:23]: Yes I understand that, i was more about having clean and uniform code here. I don't think your point as an exception. What do you think? For me the other sintax give clean and uniform ruby namespacing, and is also in sync with the rest of the code.

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 3 above)
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb]: fair point about event to_bson. But in order to be consistent with the current event implementation, I would say we get two methods, one with and one without metadata, makes sense to you?

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 4 above)
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb:37]: What do you think about adding some logging around here about this topic? clear feedback would be nice to me.

File: lib/logstash/outputs/bson/big_decimal.rb (hunk 2 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/big_decimal.rb:29]: Maybe its a bson spec thing?

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 5 above)
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb:60]: interesting, not sure why do they have it if you mark the end with a null.

File: lib/logstash/outputs/mongodb.rb [hunk 8]
```diff
@@ -38,20 +41,18 @@ def register
...
-      if @isodate
-        # the mongodb driver wants time values as a ruby Time object.
-        # set the @timestamp value of the document to a ruby Time object, then.
-        document = event.to_hash
-      else
-        document = event.to_hash.merge("@timestamp" => event["@timestamp"].to_json)
+      # Our timestamp object now has a to_bson method, using it here
+      # why not use to_hash_with_metadata?
+      # {}.merge(other) so we don't taint the event hash innards
+      document = {}.merge(event.to_hash)
```
Comment (purbon) [inline lib/logstash/outputs/mongodb.rb:48]: why using merge with the event data? would not be more clear to use `event.to_hash.clone`? I wonder about your way of thinking here? are the implications for it?

File: lib/logstash/outputs/mongodb.rb [hunk 9]
```diff
@@ -38,20 +41,18 @@ def register
...
-
     begin
-      if @isodate
-        # the mongodb driver wants time values as a ruby Time object.
-        # set the @timestamp value of the document to a ruby Time object, then.
-        document = event.to_hash
-      else
-        document = event.to_hash.merge("@timestamp" => event["@timestamp"].to_json)
+      # Our timestamp object now has a to_bson method, using it here
+      # why not use to_hash_with_metadata?
```
Comment (purbon) [inline lib/logstash/outputs/mongodb.rb]: I don't follow this comment? I guess if we change the to_bson, not to include metadata, we can fairly use it here. what do you think?

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 3 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_event.rb]: These methods are called by the Mongo ruby client not us, so here, we are define the methods of their API.  They would not call any others we define.

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 4 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_event.rb:37]: Which logger instance would we use? This code runs in the Mongo client.

File: spec/bson/big_decimal_spec.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  it "bson_type returns a binary encoded 1" do
+    expect(subject.bson_type).to eq("\x01".force_encoding(BSON::BINARY))
+  end
+
+  describe "class methods" do
+    it "builds a new BigDecimal from BSON" do
+      decoded = described_class.from_bson(StringIO.new(bson_number))
+      expect(decoded).to eq(subject)
```
Comment (purbon) [inline spec/bson/big_decimal_spec.rb]: I know this is being picky about format, but I think this are actually two test. One that the decoded value it has the right value, and another that the decoded value has the right class. I would recommend splitting into two test examples.

File: spec/bson/logstash_timestamp_spec.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  it "bson_type returns a binary encoded 9" do
+    expect(subject.bson_type).to eq("\x09".force_encoding(BSON::BINARY))
+  end
+
+  describe "class methods" do
+    it "builds a new Timestamp from BSON" do
+      decoded = described_class.from_bson(StringIO.new(bson_time))
+      expect(decoded).to eq(subject)
```
Comment (purbon) [inline spec/bson/logstash_timestamp_spec.rb]: same comment as before for big decimal.

File: spec/bson/logstash_timestamp_spec.rb [hunk 12]
```diff
@@ -0,0 +1,30 @@
+# encoding: utf-8
+require_relative "../spec_helper"
+require 'stringio'
+
+describe ::LogStash::Timestamp do
```
Comment (purbon) [inline spec/bson/logstash_timestamp_spec.rb:5]: I see a pattern for this spec, can we might be make a shared example with the other bson_type specs? what do you think?

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 6 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/logstash_event.rb:23]: It is clean and uniform - in exactly the same style as the other bson extensions in ruby-bson. If anything, the bson folder should be in lib directly, so the folder/file structure mimics the modules structure.
@purbon - This is not really something that needs changing IMO.

**Comment by purbon** [comment]
Hi, left you some comments, also I executed the tests and got this errors:

```

Failures:

  1) LogStash::Timestamp to_bson returns a binary encoded timestamp
     Failure/Error: expect(subject.to_bson).to eq(bson_time)

       expected: "\x80{y@\x88\xFE\xFF\xFF"
            got: "\x00\x8DB@\x88\xFE\xFF\xFF"

       (compared using ==)
     # ./spec/bson/logstash_timestamp_spec.rb:16:in `(root)'

  2) LogStash::Timestamp class methods builds a new Timestamp from BSON
     Failure/Error: expect(decoded).to eq(subject)

       expected: "1918-11-11T10:00:00.000Z"
            got: "1918-11-11T11:00:00.000Z"

       (compared using ==)

       Diff:
       @@ -1,2 +1,2 @@
       -"1918-11-11T10:00:00.000Z"
       +"1918-11-11T11:00:00.000Z"
     # ./spec/bson/logstash_timestamp_spec.rb:26:in `(root)'

Finished in 0.11 seconds (files took 2.32 seconds to load)
10 examples, 2 failures

Failed examples:

rspec ./spec/bson/logstash_timestamp_spec.rb:15 # LogStash::Timestamp to_bson returns a binary encoded timestamp
rspec ./spec/bson/logstash_timestamp_spec.rb:24 # LogStash::Timestamp class methods builds a new Timestamp from BSON

Randomized with seed 11911
```

what do you think it could be?

**Comment by purbon** [comment]
@guyboertje what do you think about adding code for:

`Although the test does not check the contents of what it reads back from Mongo, I did it manually in the mongo console.` 

does it makes sense to you?

File: lib/logstash/outputs/bson/logstash_event.rb (hunk 6 above)
Comment (purbon) [inline lib/logstash/outputs/bson/logstash_event.rb:23]: ok, I find this strange, but up to you. :+1:

File: lib/logstash/outputs/mongodb.rb (hunk 8 above)
Comment (guyboertje) [inline lib/logstash/outputs/mongodb.rb:48]: My thinking is this:
- `event.to_hash` is a non standard method on an object that this class knows nothing about.
- there is no compiler/interface certainty that `event.to_hash` will return a Ruby Hash, it only has to be hash-like by convention.
- there is no certainty that the object returned by `to_hash` is clonable.
- there is no certainty about how the `clone` is implemented.
- By using `{}.merge`, I am establishing three things: 1) document will be a Ruby Hash, 2) it is independent from the event, 3) the object given to merge only needs to implement `each {|k,v| ...}`.

All in all, it makes this code less dependent on the Event implementation, now and in the future.

File: lib/logstash/outputs/mongodb.rb (hunk 9 above)
Comment (guyboertje) [inline lib/logstash/outputs/mongodb.rb]: I was referring to this code comment in this [PR comment](https://github.com/logstash-plugins/logstash-output-mongodb/pull/19#issuecomment-163152344)

**Comment by guyboertje** [comment]
The failing test is timezone related.  I will add `.utc` in the spec.

**Comment by guyboertje** [comment]
Re: changing the integration test.  I don't feel strongly about changing it and I don't want it to hold up this PR any longer.  You should create an issue to expand the integration test. It would be a good one for the community who would have a better knowledge of Mongo.

File: spec/bson/big_decimal_spec.rb (hunk 10 above)
Comment (guyboertje) [inline spec/bson/big_decimal_spec.rb]: I don't agree. I feel that what you suggest is rspec overkill in this case. The right class expectation checks that `new()` surrounds the from bson conversion - this is something, that once in place is very very unlikely to change and therefore the expectation is very very likely to be met.  The right value expectation is different - it depends on what the Mongo ruby driver gives to the `from_bson` method and whether it can be decoded by the `from_bson` implementation.

But the above is moot because the test is not an integration test, we are not really getting a value from the Mongo ruby driver to use for the `from_bson` tests.

Further, the `from_bson` methods would only be used by a mongo-input plugin when ingesting data from mongo.

When we have a mongo input, these bson extensions and tests will need to move to a common place and a read integration test created in the new plugin, but a read integration test would not yield an instance of BigDecimal or Timestamp because the type encoded for BigDecimal is a Float so we would get a Float back, Timestamp is encoded as a Time type so we would get a Time back.

In conclusion, the `from_bson` methods will never be used at all.  So the tests for these methods do not mean any more than "was the code sort-of correct at the time of writing".

I can't see any reason to change the tests.

**Comment by purbon** [comment]
yeah, agree on the integration test improvement.

File: lib/logstash/outputs/bson/big_decimal.rb (hunk 1 above)
Comment (guyboertje) [inline lib/logstash/outputs/bson/big_decimal.rb:1]: I am leaving the comments in the files as is. See [Apache 2 license](http://www.apache.org/licenses/LICENSE-2.0) clause 4: Redistribution, paragraph c.

File: spec/bson/logstash_timestamp_spec.rb (hunk 12 above)
Comment (guyboertje) [inline spec/bson/logstash_timestamp_spec.rb:5]: If and when we need more of these patches, we can debate a shared example.

**Comment by guyboertje** [comment]
@purbon - the tests should now pass in EU timezone.  Any more changes?

**Comment by guyboertje** [comment]
We don't need to backport at this time.

**Comment by purbon** [comment]
LGTM
